### ISSUE TITLE
two problems about algorithm convergence

### ISSUE THREAD

**Issue by gaoxueyi0532** [tianyic/only_train_once#28]
the first problem: why the convergence process is very slow? bellow is my opt setting:
`    opt = oto.dhspg(
                    variant='sgd',
                    lr=0.01,
                    first_momentum=0.9,
                    lmbda=1e-2,
                    lmbda_amplify=20,
                    hat_lmbda_coeff=100,
                    target_group_sparsity=0.3,
                    weight_decay=1e-4,
                    start_pruning_steps=20 * len(train_loader), 
                    epsilon=0.9)`
150 epoches total, wheh finished 120 epoches, the group_sparsity only reaches to 0.14, how to improve?

the second problem, : why the final group_sparsity greater than target_group_sparsity? bellow is my opt setting:
`    opt = oto.dhspg(
                    variant='adam',
                    target_group_sparsity=0.3,
                    weight_decay=1e-4,
                    start_pruning_steps=20 * len(train_loader), 
                    epsilon=0.9)`
150 epoches total, wheh finished 120 epoches, but the final group_sparsity is 0.634(greater than 0.3), What's the problem？

hope your response, thanks a lot!

**Comment by tianyic** [comment]
The first question is good. You could try the below setup.

```python
opt = oto.dhspg(variant='sgd', 
                lr=0.01, 
                first_momentum=0.9, 
                lmbda=1e-1, 
                lmbda_amplify=100, 
                hat_lmbda_coeff=1000, 
                target_group_sparsity=0.3, 
                weight_decay=1e-4, 
                start_pruning_steps=20 * len(train_loader),  
                epsilon=0.95)
```

The second question is very strange as during our hundreds of experiments onto varying application. Group sparsity significantly higher than the target **never** happened. 

If indeed happen, please record the below six metrics and share the log back to us, which will give us more insights. 

```python 
group_sparsity, omega = optimizer.compute_group_sparsity_omega()
norm_x_pm, norm_x_npm, num_groups_pm, num_groups_npm = optimizer.compute_norm_group_partitions()
```

**Comment by gaoxueyi0532** [comment]
ok , i will supply more log information abount the second problem

**Comment by gaoxueyi0532** [comment]
The reason for the second question may be that no Learning rate is setted, bellow is my code:
`opt = oto.dhspg( variant='adam', target_group_sparsity=0.3, weight_decay=1e-4, start_pruning_steps=20 * len(train_loader),  epsilon=0.9)`

but DHSPG's __init__() is: 
`def __init__(self, params, variant='sgd', lr=required...` 

In all other experiments, lr is setted explicitly.

**Comment by tianyic** [comment]
Closed as resolved.
